port/stmicro/stm32/src/hals/STM32F103/i2c.zig

```diff
@@ -0,0 +1,362 @@
+    }
+
+    pub fn runtime_apply(i2c: I2C, config: Config) ConfigError!void {
+        const mode = config.mode;
+        const pclk = config.pclk;
+        try validade_pclk(pclk, mode);
+        try validade_speed(config.speed, mode);
+        const Trise = try validade_Trise(pclk, mode);
+        const CCR = try validade_CCR(config);
+
+        i2c.apply_internla(config, CCR, Trise);
+    }
+
+    fn apply_internla(i2c: I2C, config: Config, CCR: usize, Trise: usize) void {
+        const regs = get_regs(i2c);
+        const val: u6 = @intCast(config.pclk / 1_000_000);
+        const duty: usize = if (config.enable_duty) 1 else 0;
+        const mode: F_S = @enumFromInt(@intFromEnum(config.mode));
+
+        regs.CR1.modify(.{
+            .PE = 0,
+        });
+        i2c.reset();
+
+        regs.CR2.modify(.{
+            .FREQ = val,
+        });
+
+        regs.TRISE.modify(.{
+            .TRISE = @as(u6, @intCast(Trise)),
+        });
+
+        regs.CCR.modify(.{
+            .CCR = @as(u12, @intCast(CCR)),
+            .DUTY = @as(DUTY, @enumFromInt(duty)),
+            .F_S = mode,
+        });
+
+        regs.CR1.modify(.{
+            .PE = 1,
+        });
+    }
+
+    fn reset(i2c: I2C) void {
+        const regs = get_regs(i2c);
+        regs.CR1.modify(.{ .SWRST = 1 });
+        regs.CR1.modify(.{ .SWRST = 0 });
+    }
+
+    fn check_error(i2c: I2C) IOError!void {
+        const regs = get_regs(i2c);
+        const status = regs.SR1.read();
+        if (status.AF == 1) {
+            return IOError.ACK_FAILURE;
+        } else if (status.ARLO == 1) {
+            return IOError.ARBITRATION_LOST;
+        } else if (status.BERR == 1) {
+            return IOError.BUS_ERROR;
+        } else if (status.TIMEOUT == 1) {
+            return IOError.BUS_TIMEOUT;
+        }
+    }
+
+    pub fn clear_errors(i2c: I2C) void {
+        const regs = get_regs(i2c);
+        regs.SR1.modify(.{
+            .AF = 0,
+            .ARLO = 0,
+            .BERR = 0,
+        });
+        while (i2c.is_busy()) {
+            i2c.STOP();
+        }
+    }
+
+    fn is_busy(i2c: I2C) bool {
+        const regs = get_regs(i2c);
+        return regs.SR2.read().BUSY == 1;
+    }
+
+    fn START(i2c: I2C) IOError!void {
+        //NOTE: check busy line here?
+        const regs = get_regs(i2c);
+
+        while (i2c.is_busy()) {
+            try i2c.check_error();
+        }
+        regs.CR1.modify(.{
+            .ACK = 1,
+            .START = 1,
+        });
+        while (regs.SR1.read().START != 1) {
+            try i2c.check_error();
+        }
+    }
+
+    fn STOP(i2c: I2C) void {
+        const regs = get_regs(i2c);
+        regs.CR1.modify(.{
+            .ACK = 0,
+            .STOP = 1,
+        });
+
+        while (regs.CR1.read().STOP != 0) {}
+    }
+
+    fn send_10bits_addr(i2c: I2C, addr: u10, IO: u1) IOError!void {
+        _ = i2c;
+        _ = addr;
+        _ = IO;
+        //TODO
```

Comment:
file issue please

Reply:
Maybe a compile error here to prevent a user from accidentally using it for now.